subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy

```diff
@@ -77,6 +79,41 @@ class DefaultProviderFactoryTest extends Specification implements ProviderAssert
         zipped.get() == 'Big black cat'
     }
 
+    def "can zip two providers and use null to remove the value"() {
+        def not = providerFactory.provider { "not" }
+        def important = providerFactory.provider { "important" }
+
+        when:
+        def zipped = providerFactory.zip(not, important) { s1, s2 -> null }
+
+        then:
+        zipped instanceof Provider
+        !zipped.isPresent()
+        zipped.getOrNull() == null
+    }
+
+    def "can zip two providers and use null to remove the value, and it is live"() {
+        def reference = new AtomicReference<String>("initial")
+        def provider = providerFactory.provider(reference::get)
+        def second = providerFactory.provider { "value" }
+
+        when:
+        def zipped = providerFactory.zip(provider, second) { s1, s2 ->
+            s1 == "initial" ? "$s1 $s2" : null
+        }
+
+        then:
+        zipped instanceof Provider
+        zipped.isPresent()
+        zipped.get() == "initial value"
+
+        when:
+        reference.set("changed")
+
+        then:
+        !zipped.isPresent()
```

Comment:
Nit: add the check `zipped.getOrNull() == null` for completeness

Reply:
Added.